extra_data/keydata.py

```diff
@@ -467,33 +580,28 @@ def xarray(self, extra_dims=None, roi=(), name=None):
 
         ndarr = self.ndarray(roi=roi)
 
-        # Dimension labels
-        if extra_dims is None:
-            extra_dims = ['dim_%d' % i for i in range(ndarr.ndim - 1)]
-        dims = ['trainId'] + extra_dims
-
         # Train ID index
         coords = {'trainId': self.train_id_coordinates()}
 
-        if roi:
-            if not isinstance(roi, tuple):
-                roi = (roi,)
-
-            # Add coordinates for ROI
-            # We iterate through the slices in roi, and for each slice, we
-            # create a coordinate array for the corresponding dimension in the
-            # output. An integer in roi drops a dimension, so we only advance
-            # in the output dimensions when we see a slice.
-            out_dim_ax = 0
-            for i, dim_slice in enumerate(roi):
-                if isinstance(dim_slice, slice):
-                    if out_dim_ax < len(extra_dims):
-                        dim_name = extra_dims[out_dim_ax]
-                        start, stop, step = dim_slice.indices(self.entry_shape[i])
-                        coords[dim_name] = np.arange(start, stop, step)
-                    out_dim_ax += 1
-            for dim in range(len(roi), len(self.entry_shape)):
-                coords[extra_dims[dim]] = np.arange(self.entry_shape[dim])
+        def _get_dim_name(idx):
+            if extra_dims is not None:
+                return extra_dims[idx]
+            else:
+                return f'dim_{idx}'
+
+        # Dimension labels after the train dimension
+        _extra_dims = []
+        if roi != ():
+            for idx, coord in enumerate(expand_indexing(self.entry_shape, roi, expand_full=False)):
+                dim = _get_dim_name(idx)
+
+                if coord is None or not isinstance(coord, int):
+                    _extra_dims.append(dim)
+                if coord is not None and not isinstance(coord, int):
+                    coords[dim] = coord
```

Comment:
I thought from your description that you did want to keep the scalar coordinate label when a dimension is indexed with an integer, whereas this discards it. Did I misunderstand?

Reply:
Is it possible to add coordinates to a non existing dimension?

Reply:
Yup! Xarray keeps the coordinate when you select a scalar on that dimension, even though it's no longer in `dims`, so  `data.sel(x=10).coords['x']` gives you a 0D array containing the number 10.

Reply:
oh OK, it works but only with a scalar :bulb: